### ISSUE TITLE
The cuda extension is never used for 3090s?

### ISSUE THREAD

**Issue by Ph0rk0z** [thu-ml/SageAttention#77]
I did some experimenting. I got the cuda extension to build on turning by ifdef all of the BF16 stuff. The extension for SM86 can build on cuda 11.8, but it seems when I built for SM_75 it would crash unless I used cuda12. Examining the core file led me to another conclusion. For SM86, the cuda extension is never used. One would have to manually call it otherwise you get triton (naturally works for SM75).

The difference I see between 1.06 and 2.01 triton is that on 3090/2080 it is slightly slower, but the outputs are less likely to change. It's still faster than xformers. Literal fractions of a second for the same model/prompt/etc. Cuda is a little faster.

anyways, if anyone else wants to try on SM_75: https://github.com/Ph0rk0z/SageAttention2/commit/2e8133672d96b2cf48e63ed7c99a431578d7d018

**Comment by jason-huang03** [comment]
@Ph0rk0z 
Hi, the `sageattn` routes input to different kernels automatically and we believe that the triton kernel is suitable for 3090 GPU which is SM86. However, users are welcome to experiment with other kernels if they are looking for different trade-offs in speed or accuracy.

By the way you say `1.06 and 2.01 triton is that on 3090/2080 it is slightly slower`. Do you mean 1.0.6 triton is slower or 2.0.1 triton is slower? By how much?

**Comment by Ph0rk0z** [comment]
It takes 2.0.1 with fused kernel to match same speed as 1.0.6 in triton alone. The difference isn't huge. My simple workflow goes from 4.75s to 5.0 seconds between the versions. Also on some SDXL outputs, the people grow extra limbs in sageattention that aren't present in xformers using the same exact seed and all. Must be a result of the int8 quanting. No free lunch.

I will test some more with bigger models like cogvideo since maybe there are gains at the top, rather than for smaller resolutions. I'm sure most people not using this for SD/flux, etc and rather the video models.

**Comment by jason-huang03** [comment]
Yes our approach can achieve significant speedup on model that has long sequences like CogvideoX, Mochi and HuyuanDiT.
